contrib/adaptive-compression/adapt.c

```diff
@@ -0,0 +1,591 @@
+        int const fileCloseError =  (ctx->dstFile != NULL && ctx->dstFile != stdout) ? fclose(ctx->dstFile) : 0;
+        int const cctxError = ZSTD_isError(ZSTD_freeCCtx(ctx->cctx)) ? 1 : 0;
+        if (ctx->jobs){
+            freeCompressionJobs(ctx);
+            free(ctx->jobs);
+        }
+        return compressedMutexError | compressedCondError | readyMutexError | readyCondError | fileCloseError | allJobsMutexError | allJobsCondError | jobWriteMutexError | jobWriteCondError | cctxError;
+    }
+}
+
+static adaptCCtx* createCCtx(unsigned numJobs, const char* const outFilename)
+{
+
+    adaptCCtx* ctx = malloc(sizeof(adaptCCtx));
+    if (ctx == NULL) {
+        DISPLAY("Error: could not allocate space for context\n");
+        return NULL;
+    }
+    memset(ctx, 0, sizeof(adaptCCtx));
+    ctx->compressionLevel = g_compressionLevel;
+    pthread_mutex_init(&ctx->jobCompressed_mutex, NULL); 
+    pthread_cond_init(&ctx->jobCompressed_cond, NULL);
+    pthread_mutex_init(&ctx->jobReady_mutex, NULL);
+    pthread_cond_init(&ctx->jobReady_cond, NULL);
+    pthread_mutex_init(&ctx->allJobsCompleted_mutex, NULL);
+    pthread_cond_init(&ctx->allJobsCompleted_cond, NULL);
+    pthread_mutex_init(&ctx->jobWrite_mutex, NULL);
+    pthread_cond_init(&ctx->jobWrite_cond, NULL);
+    ctx->numJobs = numJobs;
+    ctx->jobReadyID = 0;
+    ctx->jobCompressedID = 0;
+    ctx->jobWriteID = 0;
+    ctx->lastJobID = -1; /* intentional underflow */
+    ctx->jobs = calloc(1, numJobs*sizeof(jobDescription));
+    ctx->nextJobID = 0;
+    ctx->threadError = 0;
+    ctx->allJobsCompleted = 0;
+    ctx->adaptParam = DEFAULT_ADAPT_PARAM;
+    ctx->cctx = ZSTD_createCCtx();
+    if (!ctx->cctx) {
+        DISPLAY("Error: could not allocate ZSTD_CCtx\n");
+        freeCCtx(ctx);
+        return NULL;
+    }
+    if (!ctx->jobs) {
+        DISPLAY("Error: could not allocate space for jobs during context creation\n");
+        freeCCtx(ctx);
+        return NULL;
+    }
+    {
+        unsigned const stdoutUsed = !strcmp(outFilename, stdoutmark);
+        FILE* dstFile = stdoutUsed ? stdout : fopen(outFilename, "wb");
+        if (dstFile == NULL) {
+            DISPLAY("Error: could not open output file\n");
+            freeCCtx(ctx);
+            return NULL;
+        }
+        ctx->dstFile = dstFile;
+    }
+    return ctx;
+}
+
+
+
+static void waitUntilAllJobsCompleted(adaptCCtx* ctx)
+{
+    pthread_mutex_lock(&ctx->allJobsCompleted_mutex);
+    while (ctx->allJobsCompleted == 0) {
+        pthread_cond_wait(&ctx->allJobsCompleted_cond, &ctx->allJobsCompleted_mutex);
+    }
+    pthread_mutex_unlock(&ctx->allJobsCompleted_mutex);
+}
+
+static unsigned adaptCompressionLevel(adaptCCtx* ctx)
+{
+    unsigned reset = 0;
+    unsigned const allSlow = ctx->adaptParam < ctx->stats.compressedCounter && ctx->adaptParam < ctx->stats.writeCounter && ctx->adaptParam < ctx->stats.readyCounter ? 1 : 0;
```

Comment:
If the ternary operator results are `1` and `0`, you can just use the condition.